src/eredis_cluster.erl

```diff
@@ -377,26 +404,181 @@ query_noreply(Command, PoolKey) ->
+    filter_out_asking_results(Commands, Results, Acc);
+filter_out_asking_results([_Command | Commands], [Result | Results], Acc) ->
+    filter_out_asking_results(Commands, Results, [Result | Acc]);
+filter_out_asking_results([], [], Acc) ->
+    lists:reverse(Acc);
+filter_out_asking_results(_Commands, _Results, _Acc) ->
+    %% Length of commands and length of results mismatch
+    {error, redirect_failed}.
+
+%% Adds ASKING before every command, except inside a MULTI-command.
+add_asking_to_pipeline_command([Command|Commands]) ->
+    case string:uppercase(iolist_to_binary(Command)) of
+        <<"MULTI">> ->
+            {Transaction, AfterTransaction} = lists:splitwith(
+                                                fun is_not_exec_or_discard/1,
+                                                Commands),
+            [[<<"ASKING">>], Command | Transaction] ++
+                add_asking_to_pipeline_command(AfterTransaction);
+        _NotMulti ->
+            [[<<"ASKING">>], Command | add_asking_to_pipeline_command(Commands)]
+    end;
+add_asking_to_pipeline_command([]) ->
+    [].
+
+is_not_exec_or_discard([Command]) ->
+    case string:uppercase(iolist_to_binary(Command)) of
+        <<"EXEC">> -> false;
+        <<"DISCARD">> -> false;
+        _Other -> true
+    end;
+is_not_exec_or_discard(_Command) ->
+    true.
+
+%% Parses the Rest as in <<"ASK ", Rest/binary>> and returns an
+%% existing pool if any or an error.
+-spec parse_redirect_info(RedirectInfo :: binary()) ->
+          {ok, ExistingPool :: atom()} | {error, any()}.
+parse_redirect_info(RedirectInfo) ->
+    try
+        [_Slot, AddrPort] = binary:split(RedirectInfo, <<" ">>),
+        [Addr0, PortBin] = binary:split(AddrPort, <<":">>),
+        Port = binary_to_integer(PortBin),
+        Addr = case Addr0 of
+                   <<"[", IPv6:(byte_size(Addr0) - 2)/binary, "]">> ->
+                       %% An IPv6 address wrapped in square brackets.
+                       IPv6;
+                   _ ->
+                       Addr0
+               end,
+        eredis_cluster_pool:get_existing_pool(Addr, Port)
+    of
+        {ok, Pool} ->
+            {ok, Pool};
+        {error, _} = Error ->
+            Error
+    catch
+        error:{badmatch, _} ->
+            %% Couldn't parse using binary:split/2.
+            {error, bad_redirect};
+        error:badarg ->
+            %% binary_to_integer/1 failed
+            {error, bad_redirect}
+    end.
+
+handle_transaction_result(Results, Version) when is_list(Results) ->
+    %% Consider all errors, to make sure slot mapping is updated if
+    %% needed. (Multiple slot mapping updates have no effect if the
+    %% Version is the same.)
+    HandledResults = [handle_transaction_result(Result, Version)
+                      || Result <- Results],
+    case lists:member(retry, HandledResults) of
+        true  -> retry;
+        false -> Results
```

Comment:
would this change (or correct) the behavior a bit? we now return a list of errors instead of just one of them.
Before it was returning an error by doing a proplist lookup with the first error in list.
Maybe that info should be added to the README.md togheter with the info about the new ASK support

Reply:
Yes, in the case of `qmn/1`, it's a correction. It would crash if `handle_transaction_result` returns an error tuple as Res in this code:

```Erlang
qmn2([{Pool, PoolCommands} | T1], [{Pool, Mapping} | T2], Acc, Version) ->
    Transaction = fun(Worker) -> qw(Worker, PoolCommands) end,
    Result = eredis_cluster_pool:transaction(Pool, Transaction),
    case handle_transaction_result(Result, Version) of
        retry -> retry;
        Res ->
            MappedRes = lists:zip(Mapping, Res),
            qmn2(T1, T2, MappedRes ++ Acc, Version)
    end;
```

... and for query/3 (used by q/1 and qk/2), there was a crash (case clause) if `handle_transaction_result/2` would return anything other than `retry` or the result given as input to it:

```Erlang
query(Transaction, Slot, Counter) ->
    throttle_retries(Counter),

    {Pool, Version} = eredis_cluster_monitor:get_pool_by_slot(Slot),

    Result = eredis_cluster_pool:transaction(Pool, Transaction),
    case handle_transaction_result(Result, Version) of
        retry -> query(Transaction, Slot, Counter + 1);
        Result -> Result
    end.
```

So yes, I would say this corrects the behavior in some cases. For `qa/1` there was no problem but the return value changes slightly. Is it safe to assume that nobody depends on the old behavior of `qa/1` (given that `q/1` with a pipeline command would crash)?